timesync/peersync/sync.go

```diff
@@ -0,0 +1,330 @@
+func WithLog(lg log.Log) Option {
+	return func(s *Sync) {
+		s.log = lg
+	}
+}
+
+// WithConfig modifies config used in Sync.
+func WithConfig(config Config) Option {
+	return func(s *Sync) {
+		s.config = config
+	}
+}
+
+// New creates Sync instance and returns pointer.
+func New(network Network, opts ...Option) *Sync {
+	sync := &Sync{
+		log:     log.NewNop(),
+		ctx:     context.Background(),
+		time:    systemTime{},
+		config:  DefaultConfig(),
+		network: network,
+		peersWatcher: peersWatcher{
+			requests: make(chan *waitPeersReq),
+		},
+	}
+	for _, opt := range opts {
+		opt(sync)
+	}
+
+	sync.ctx, sync.cancel = context.WithCancel(sync.ctx)
+	sync.wg, sync.ctx = errgroup.WithContext(sync.ctx)
+
+	sync.srv = server.NewMsgServer(sync.ctx,
+		network,
+		protocolName,
+		sync.config.RoundTimeout,
+		make(chan service.DirectMessage, sync.config.ResponsesBufferSize),
+		sync.log,
+	)
+	sync.srv.RegisterBytesMsgHandler(server.RequestTimeSync, sync.requestHandler)
+	return sync
+}
+
+// Sync manages background worker that compares peers time with system time.
+type Sync struct {
+	errCnt uint32
+
+	config       Config
+	log          log.Log
+	srv          *server.MessageServer
+	time         Time
+	network      Network
+	peersWatcher peersWatcher
+
+	once   sync.Once
+	wg     *errgroup.Group
+	ctx    context.Context
+	cancel func()
+}
+
+func (s *Sync) requestHandler(ctx context.Context, buf []byte) []byte {
+	var request Request
+	if err := types.BytesToInterface(buf, &request); err != nil {
+		s.log.Debug("can't decode request", log.Binary("request", buf), log.Err(err))
+		return nil
+	}
+	resp := Response{
+		ID:        request.ID,
+		Timestamp: s.time.Now().UnixNano(),
+	}
+	buf, err := types.InterfaceToBytes(&resp)
+	if err != nil {
+		s.log.Panic("can't encode response", log.Binary("response", buf), log.Err(err))
+	}
+	return buf
+}
+
+// Start background workers.
+func (s *Sync) Start() {
+	s.once.Do(func() {
+		s.wg.Go(func() error {
+			return s.run()
+		})
+		s.wg.Go(func() error {
+			added, expired := s.network.SubscribePeerEvents()
+			return s.peersWatcher.run(s.ctx, added, expired)
+		})
+	})
+}
+
+// Stop background workers.
+func (s *Sync) Stop() {
+	s.cancel()
+	s.srv.Close()
+	s.Wait()
+}
+
+// Wait will return first error that is returned by background workers.
+func (s *Sync) Wait() error {
+	err := s.wg.Wait()
+	if errors.Is(err, context.Canceled) {
+		return nil
+	}
+	return err
+}
+
+func (s *Sync) run() error {
+	var (
+		timer *time.Timer
+		round uint64
+	)
+	s.log.With().Debug("started sync background worker")
+	defer s.log.With().Debug("exiting sync background worker")
+	for {
+		prs, err := s.peersWatcher.waitPeers(s.ctx, s.config.RequiredResponses)
+		if err != nil {
+			return err
+		}
+		s.log.With().Info("starting time sync round with peers",
+			log.Uint64("round", round),
+			log.Int("peers_count", len(prs)),
+			log.Uint32("errors_count", atomic.LoadUint32(&s.errCnt)),
```

Comment:
why does this need to be atomic? is it run in parallel?

Reply:
the idea is that the user can reset this counter after seeing a notification (or error in logs, depending on how they run a node). using an API call